client/finality-grandpa/src/environment.rs

```diff
@@ -924,22 +929,133 @@ where
 	fn prevote_equivocation(
 		&self,
 		_round: RoundNumber,
-		equivocation: ::finality_grandpa::Equivocation<Self::Id, Prevote<Block>, Self::Signature>
+		equivocation: finality_grandpa::Equivocation<Self::Id, Prevote<Block>, Self::Signature>,
 	) {
 		warn!(target: "afg", "Detected prevote equivocation in the finality worker: {:?}", equivocation);
-		// nothing yet; this could craft misbehavior reports of some kind.
+		match report_equivocation(
+			&self.client.clone(),
+			&self.authority_set.inner().read(),
+			&self.select_chain,
+			equivocation.into(),
+		) {
+			Ok(_) => {}
+			Err(err) => {
+				warn!(target: "afg", "Error reporting prevote equivocation: {:?}", err);
+			}
+		};
 	}
 
 	fn precommit_equivocation(
 		&self,
 		_round: RoundNumber,
-		equivocation: Equivocation<Self::Id, Precommit<Block>, Self::Signature>
+		equivocation: finality_grandpa::Equivocation<Self::Id, Precommit<Block>, Self::Signature>,
 	) {
 		warn!(target: "afg", "Detected precommit equivocation in the finality worker: {:?}", equivocation);
-		// nothing yet
+		match report_equivocation(
+			&self.client.clone(),
+			&self.authority_set.inner().read(),
+			&self.select_chain,
+			equivocation.into(),
+		) {
+			Ok(_) => {}
+			Err(err) => {
+				warn!(target: "afg", "Error reporting precommit equivocation: {:?}", err);
+			}
+		};
 	}
 }
 
+/// Report the given equivocation to the GRANDPA runtime module. This method
+/// generates a session membership proof of the offender and then submits an
+/// extrinsic to report the equivocation. In particular, the session membership
+/// proof must be generated at the block at which the given set was active which
+/// isn't necessarily the best block if there are pending authority set changes.
```

Comment:
when and how do we have pending authorities? is this a reflection of the queued validators in session module?

Reply:
@kianenigma From the runtime perspective, we don't have pending authorities. However, GRANDPA set changes are pending until finalized. When finality lags we can have a DAG of unfinalized changes. In normal circumstances a GRANDPA set-change is not pending for more than a few seconds.